designs/0032-stan-output-tech-spec.md

```diff
@@ -0,0 +1,74 @@
+## Technical Specification
+
+To accomplish the above require changes to parts of the `stan::services` layer.
+We will add algorithm-specific output handler objects to `stan::services`.
+which manage multiple writer callback objects e.g., `hmc_output_handler`, `advi_output_handler`, `optimization_output_handler`.
+
+These handler objects will be instantiated by the Stan interfaces.
+The interface will determine which information is output and where it is sent to.
+through use of a mapping from output types to `stan::callbacks::writer` objects.
+We will add new signatures to the `stan::services` layer calling methods
+to pass these output handlers to the inference algorithms, replacing the current slots for writer callbacks with
+a single slot for the output handler.
+This will allow for new output handlers, while still allowing for use of the
+current output formats for backwards compatibility.
+
+To decouple the output format from the output writer, we will use formatter callbacks.
+These classes will provide functions like `write_draws`, `write_unconstrained_params`
+which route the information from the inference engine or model to the appropriate callback writer
+using the appropriate formatter.
+
+We will expand the set of data types that the `stan::callbacks::writer` object can handle to include:
+
++ a vector of ints
++ Eigen array, matrix, and vector types
+
+
+### Creating Apache Arrow schemas for inference algorithm and Stan model data
+
+The instantiated model object provides the information used to create an Apache Arrow schema
+used to serialize the values produced by the model's `write_array` function.
+### Stan program outputs
+
+The Stan model class `write_array` method populates a vector of doubles with the values for all
```

Comment:
How does this vector of doubles square with the above description of being able to accept Eigen types and integers in the writer interfaces?

Reply:
for `write_array` some of those doubles are actually representing ints.
some outputs from algorithms might return Eigen types.
I think we should design for this eventuality.

Reply:
Yes, it's difficult to think of a way in which write_array could be written generically. It would be great if we could though, since we could keep the high-level abstraction without needing as many packing/unpacking loops

Reply:
changes to write_array are way beyond the scope of this proposal.

Reply:
So, the overloads discussed in this tech spec (i.e. writer callbacks which can accept a vector of ints, or an eigen matrix) are not for model output but rather for internals to the algorithms or that sort of thing?